### PR TITLE
add elasticlunrjs definitions

### PR DESCRIPTION
**Pull request by robertIsaac**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Review by elesueur — APPROVED** [review]
Looks good Rob.

BTW, how does this get versioned?

**Comment by robertIsaac** [comment]
> Looks good Rob.
> 
> BTW, how does this get versioned?

I write it manually as 1.0 to match the npm version https://www.npmjs.com/package/elasticlunrjs

File: types/elasticlunrjs/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,386 @@
+// Type definitions for elasticlunrjs 1.0
+// Project: https://github.com/robertIsaac/elasticlunr.js
+// Definitions by: Robert Isaac <https://github.com/robertIsaac>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export = elasticlunr;
```
Comment (peterblazejewicz) [inline types/elasticlunrjs/index.d.ts:6]:
```suggestion
export = elasticlunr;

export as namespace elasticlunr;
```
this is most probably a proper UMD module:
https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#identifying-a-umd-library

File: types/elasticlunrjs/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+        static load<T>(serialisedData: SerialisedDocumentStore<T>): DocumentStore<T>;
+    }
+
+    type EventType = 'add' | 'update' | 'remove';
+    type EventHandler = (...args: any[]) => void;
+
+    class EventEmitter {
+        constructor();
+
```
Comment (peterblazejewicz) [inline types/elasticlunrjs/index.d.ts:58]:
```suggestion
```
all classes ships with parameterless cnstructors

File: types/elasticlunrjs/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        fields: Array<keyof T>;
+        ref: keyof T;
+        pipeline: SerialisedPipeline;
+        documentStore: SerialisedDocumentStore<T>;
+        index: { [K in keyof T]?: InvertedIndexNode };
+    }
+
+    class Index<T extends {}> {
+        constructor();
+
```
Comment (peterblazejewicz) [inline types/elasticlunrjs/index.d.ts:128]:
```suggestion
```
ditto

File: types/elasticlunrjs/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    }
+
+    type InvertedIndexNode = InvertedIndexCharNode & {
+        df: number;
+        docs: InvertedIndexDocs;
+    };
+
+    class InvertedIndex {
+        constructor();
+
```
Comment (peterblazejewicz) [inline types/elasticlunrjs/index.d.ts:225]:
```suggestion
```

File: types/elasticlunrjs/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+        static load(serialisedData: SerialisedInvertedIndex): InvertedIndex;
+    }
+
+    type SerialisedPipeline = string[];
+
+    type PipelineFunction = (token: string, i: number, tokens: string[]) => string | undefined | null | void;
+
+    class Pipeline {
+        constructor();
+
```
Comment (peterblazejewicz) [inline types/elasticlunrjs/index.d.ts:253]:
```suggestion
```

File: types/elasticlunrjs/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        static load(serialised: SerialisedPipeline): Pipeline;
+
+        static registerFunction(fn: PipelineFunction, label: string): void;
+
+        static warnIfFunctionNotRegistered(fn: PipelineFunction): void;
+    }
+
+    class SortedSet<T> {
+        constructor();
+
```
Comment (peterblazejewicz) [inline types/elasticlunrjs/index.d.ts:281]:
```suggestion
```

**Comment by peterblazejewicz** [comment]
are the params like `i` or `w` taken literally from JS module? those can have more context for users I  think.

**Review by elesueur — APPROVED** [review]
LG

**Comment by peterblazejewicz** [comment]
> are the params like `i` or `w` taken literally from JS module? those can have more context for users I think.

what about those? are those `index`, `word`? I rely a lot on params (and IDE providing them, unless using copilot)

**Comment by robertIsaac** [comment]
> are the params like `i` or `w` taken literally from JS module? those can have more context for users I think.

they are not an explicit params

**Comment by robertIsaac** [comment]
> > are the params like `i` or `w` taken literally from JS module? those can have more context for users I think.
> 
> what about those? are those `index`, `word`? I rely a lot on params (and IDE providing them, unless using copilot)

actually the w is token, will push it now
for the I'm still looking

**Comment by robertIsaac** [comment]
> > > are the params like `i` or `w` taken literally from JS module? those can have more context for users I think.
> > 
> > 
> > what about those? are those `index`, `word`? I rely a lot on params (and IDE providing them, unless using copilot)
> 
> actually the `w` is token, will push it now for the `i` I'm still looking

i is the token index

**Review by peterblazejewicz — APPROVED** [review]
LGTM!

**Comment by robertIsaac** [comment]
Ready to merge

**Pull request merged**